File: src/test/java/tool/cmnclslib/mdl/UnitTest_MdlNet.java
package tool.cmnclslib.mdl;

import java.net.InetAddress;
import java.net.ServerSocket;
import java.util.List;
import org.junit.Test;
import static org.junit.Assert.*;

/**
 * MdlNet の単体テストクラスです。
 */
public class UnitTest_MdlNet {

    @Test
    public void icmpPing_HostString_Loopback() {
        boolean result = MdlNet.icmpPing("127.0.0.1", 0, 2000);
        assertTrue(result);
    }

    @Test
    public void icmpPing_HostString_InvalidHost_ReturnsFalse() {
        boolean result = MdlNet.icmpPing("invalid_host_name_999999", 1, 500);
        assertFalse(result);
    }

    @Test
    public void icmpPingAsync_HostString_Loopback() throws Exception {
        boolean result = MdlNet.icmpPingAsync("127.0.0.1", 2, 2000).get();
        assertTrue(result);
    }

    @Test
    public void icmpPingAsync_HostString_InvalidHost_ReturnsFalse() throws Exception {
        boolean result = MdlNet.icmpPingAsync("invalid_host_name_999999", 0, 500).get();
        assertFalse(result);
    }

    @Test
    public void icmpPing_InetAddress_Loopback() throws Exception {
        boolean result = MdlNet.icmpPing(InetAddress.getByName("127.0.0.1"), 1, 2000);
        assertTrue(result);
    }

    @Test
    public void icmpPingAsync_InetAddress_Loopback() throws Exception {
        boolean result = MdlNet.icmpPingAsync(InetAddress.getByName("127.0.0.1"), 1, 2000).get();
        assertTrue(result);
    }

    @Test
    public void tcpPing_HostString_SuccessAndFailure() throws Exception {
        int port;
        try (ServerSocket serverSocket = new ServerSocket(0, 1, InetAddress.getByName("127.0.0.1"))) {
            port = serverSocket.getLocalPort();
            boolean success = MdlNet.tcpPing("127.0.0.1", port, 1, 2000);
            assertTrue(success);
        }

        boolean fail = MdlNet.tcpPing("127.0.0.1", port, 1, 500);
        assertFalse(fail);
    }

    @Test
    public void tcpPingAsync_HostString_SuccessAndFailure() throws Exception {
        int port;
        try (ServerSocket serverSocket = new ServerSocket(0, 1, InetAddress.getByName("127.0.0.1"))) {
            port = serverSocket.getLocalPort();
            boolean success = MdlNet.tcpPingAsync("127.0.0.1", port, 0, 2000).get();
            assertTrue(success);
        }

        boolean fail = MdlNet.tcpPingAsync("127.0.0.1", port, 1, 500).get();
        assertFalse(fail);
    }

    @Test
    public void getHostnameFromIpAddress_Loopback_ReturnsHostnameOrIp() {
        String hostname = MdlNet.getHostnameFromIpAddress("127.0.0.1");
        assertFalse(hostname.isEmpty());
    }

    @Test
    public void getHostnameFromIpAddress_InvalidIp_ReturnsOriginalString() {
        String invalidIp = "999.999.999.999";
        String hostname = MdlNet.getHostnameFromIpAddress(invalidIp);
        assertEquals(invalidIp, hostname);
    }

    @Test
    public void getIpAddressFromHostname_Localhost_IPv4() {
        InetAddress ip = MdlNet.getIpAddressFromHostname("localhost", false);
        assertNotNull(ip);
    }

    @Test
    public void getIpAddressFromHostname_InvalidHost_ReturnsNull() {
        InetAddress ip = MdlNet.getIpAddressFromHostname("invalid_host_name_999999", false);
        assertNull(ip);
    }

    @Test
    public void getIpStringFromHostname_Localhost() {
        String ipStr = MdlNet.getIpStringFromHostname("localhost", false);
        assertNotNull(ipStr);
    }

    @Test
    public void getIpStringFromHostname_NullOrEmptyHost_ReturnsNull() {
        String ipStr1 = MdlNet.getIpStringFromHostname(null, false);
        String ipStr2 = MdlNet.getIpStringFromHostname("", false);
        assertNull(ipStr1);
        assertNull(ipStr2);
    }

    @Test
    public void getIpv4Addresses_ReturnsList() {
        List<String> addressesIncludeDown = MdlNet.getIpv4Addresses(true);
        List<String> addressesActiveOnly = MdlNet.getIpv4Addresses(false);

        assertNotNull(addressesIncludeDown);
        assertNotNull(addressesActiveOnly);
    }

    @Test
    public void 新設メソッド_getHostFromIp_getIpFromHost等が動作すること() {
        assertEquals("127.0.0.1", MdlNet.getHostFromIp("127.0.0.1"));
        assertNotNull(MdlNet.getHostFromIpAsync("127.0.0.1"));
        assertNotNull(MdlNet.getIpFromHost("localhost", false));
        assertNotNull(MdlNet.getIpFromHostAsync("localhost", false));
        assertNotNull(MdlNet.getIpStrFromHost("localhost", false));
        assertNotNull(MdlNet.getIpStrHostAsync("localhost", false));
    }
}

File: TestProject1/Module/UnitTest_MdlNet.cs
using System;
using System.Collections.Generic;
using System.Net;
using System.Net.Sockets;
using System.Threading;
using System.Threading.Tasks;
using CmnClsLib.Module;
using Xunit;
using Assert = Xunit.Assert;

namespace TestProject1.Module
{
    public class UnitTest_MdlNet
    {
        #region ICMP Ping Tests

        [Theory]
        [InlineData("127.0.0.1", 0)]
        [InlineData("127.0.0.1", 1)]
        [InlineData("127.0.0.1", 2)]
        public void IcmpPing_HostString_Loopback(string host, int verbosity)
        {
            bool result = MdlNet.IcmpPing(host, verbosity, timeout: 2000);
            Assert.True(result);
        }

        [Fact]
        public void IcmpPing_HostString_InvalidHost_ReturnsFalse()
        {
            bool result = MdlNet.IcmpPing("invalid_host_name_999999", verbosity: 1, timeout: 500);
            Assert.False(result);
        }

        [Fact]
        public async Task IcmpPingAsync_HostString_Loopback()
        {
            bool result = await MdlNet.IcmpPingAsync("127.0.0.1", verbosity: 2, timeout: 2000);
            Assert.True(result);
        }

        [Fact]
        public async Task IcmpPingAsync_HostString_InvalidHost_ReturnsFalse()
        {
            bool result = await MdlNet.IcmpPingAsync("invalid_host_name_999999", verbosity: 0, timeout: 500);
            Assert.False(result);
        }

        [Fact]
        public void IcmpPing_IPAddress_Loopback()
        {
            bool result = MdlNet.IcmpPing(IPAddress.Loopback, verbosity: 1, timeout: 2000);
            Assert.True(result);
        }

        [Fact]
        public async Task IcmpPingAsync_IPAddress_Loopback()
        {
            bool result = await MdlNet.IcmpPingAsync(IPAddress.Loopback, verbosity: 1, timeout: 2000);
            Assert.True(result);
        }

        #endregion

        #region TCP Ping Tests

        [Fact]
        public void TcpPing_HostString_SuccessAndFailure()
        {
            // 有効なローカルTCPリスナーを起動してテスト
            var listener = new TcpListener(IPAddress.Loopback, 0);
            listener.Start();
            int port = ((IPEndPoint)listener.LocalEndpoint).Port;

            try
            {
                bool success = MdlNet.TcpPing("127.0.0.1", port, verbosity: 1, timeout: 2000);
                Assert.True(success);
            }
            finally
            {
                listener.Stop();
            }

            // 停止後のポートへ接続（失敗するはず）
            bool fail = MdlNet.TcpPing("127.0.0.1", port, verbosity: 1, timeout: 500);
            Assert.False(fail);
        }

        [Fact]
        public async Task TcpPingAsync_HostString_SuccessAndFailure()
        {
            var listener = new TcpListener(IPAddress.Loopback, 0);
            listener.Start();
            int port = ((IPEndPoint)listener.LocalEndpoint).Port;

            try
            {
                bool success = await MdlNet.TcpPingAsync("127.0.0.1", port, verbosity: 0, timeout: 2000);
                Assert.True(success);
            }
            finally
            {
                listener.Stop();
            }

            bool fail = await MdlNet.TcpPingAsync("127.0.0.1", port, verbosity: 1, timeout: 500);
            Assert.False(fail);
        }

        [Fact]
        public void TcpPing_IPAddress_SuccessAndFailure()
        {
            var listener = new TcpListener(IPAddress.Loopback, 0);
            listener.Start();
            int port = ((IPEndPoint)listener.LocalEndpoint).Port;

            try
            {
                bool success = MdlNet.TcpPing(IPAddress.Loopback, port, verbosity: 1, timeout: 2000);
                Assert.True(success);
            }
            finally
            {
                listener.Stop();
            }

            bool fail = MdlNet.TcpPing(IPAddress.Loopback, port, verbosity: 1, timeout: 500);
            Assert.False(fail);
        }

        [Fact]
        public async Task TcpPingAsync_IPAddress_SuccessAndFailure()
        {
            var listener = new TcpListener(IPAddress.Loopback, 0);
            listener.Start();
            int port = ((IPEndPoint)listener.LocalEndpoint).Port;

            try
            {
                bool success = await MdlNet.TcpPingAsync(IPAddress.Loopback, port, verbosity: 1, timeout: 2000);
                Assert.True(success);
            }
            finally
            {
                listener.Stop();
            }

            bool fail = await MdlNet.TcpPingAsync(IPAddress.Loopback, port, verbosity: 1, timeout: 500);
            Assert.False(fail);
        }

        #endregion

        #region DNS / IP Name Resolution Tests

        [Fact]
        public void GetHostnameFromIpAddress_Loopback_ReturnsHostnameOrIp()
        {
            string hostname = MdlNet.GetHostnameFromIpAddress("127.0.0.1");
            Assert.False(string.IsNullOrEmpty(hostname));
        }

        [Fact]
        public void GetHostnameFromIpAddress_InvalidIp_ReturnsOriginalString()
        {
            string invalidIp = "999.999.999.999";
            string hostname = MdlNet.GetHostnameFromIpAddress(invalidIp);
            Assert.Equal(invalidIp, hostname);
        }

        [Fact]
        public async Task GetHostnameFromIpAddressAsync_Loopback_ReturnsHostnameOrIp()
        {
            string hostname = await MdlNet.GetHostnameFromIpAddressAsync("127.0.0.1");
            Assert.False(string.IsNullOrEmpty(hostname));
        }

        [Fact]
        public async Task GetHostnameFromIpAddressAsync_InvalidIp_ReturnsOriginalString()
        {
            string invalidIp = "999.999.999.999";
            string hostname = await MdlNet.GetHostnameFromIpAddressAsync(invalidIp);
            Assert.Equal(invalidIp, hostname);
        }

        [Fact]
        public void GetIpAddressFromHostname_Localhost_IPv4()
        {
            IPAddress? ip = MdlNet.GetIpAddressFromHostname("localhost", isIpv6: false);
            Assert.NotNull(ip);
            Assert.Equal(AddressFamily.InterNetwork, ip.AddressFamily);
        }

        [Fact]
        public void GetIpAddressFromHostname_InvalidHost_ReturnsNull()
        {
            IPAddress? ip = MdlNet.GetIpAddressFromHostname("invalid_host_name_999999", isIpv6: false);
            Assert.Null(ip);
        }

        [Fact]
        public async Task GetIpAddressFromHostnameAsync_Localhost_IPv4()
        {
            IPAddress? ip = await MdlNet.GetIpAddressFromHostnameAsync("localhost", isIpv6: false);
            Assert.NotNull(ip);
            Assert.Equal(AddressFamily.InterNetwork, ip.AddressFamily);
        }

        [Fact]
        public async Task GetIpAddressFromHostnameAsync_InvalidHost_ReturnsNull()
        {
            IPAddress? ip = await MdlNet.GetIpAddressFromHostnameAsync("invalid_host_name_999999", isIpv6: false);
            Assert.Null(ip);
        }

        [Fact]
        public void GetIpStringFromHostname_Localhost()
        {
            string? ipStr = MdlNet.GetIpStringFromHostname("localhost", isIpv6: false);
            Assert.NotNull(ipStr);
            Assert.True(IPAddress.TryParse(ipStr, out _));
        }

        [Fact]
        public void GetIpStringFromHostname_NullOrEmptyHost_ReturnsNull()
        {
            string? ipStr1 = MdlNet.GetIpStringFromHostname(null!, isIpv6: false);
            string? ipStr2 = MdlNet.GetIpStringFromHostname("", isIpv6: false);
            Assert.Null(ipStr1);
            Assert.Null(ipStr2);
        }

        [Fact]
        public async Task GetIpStringFromHostnameAsync_Localhost()
        {
            string? ipStr = await MdlNet.GetIpStringFromHostnameAsync("localhost", isIpv6: false);
            Assert.NotNull(ipStr);
            Assert.True(IPAddress.TryParse(ipStr, out _));
        }

        [Fact]
        public async Task GetIpStringFromHostnameAsync_NullOrEmptyHost_ReturnsNull()
        {
            string? ipStr1 = await MdlNet.GetIpStringFromHostnameAsync(null!, isIpv6: false);
            string? ipStr2 = await MdlNet.GetIpStringFromHostnameAsync("", isIpv6: false);
            Assert.Null(ipStr1);
            Assert.Null(ipStr2);
        }

        #endregion

        #region Network Interface Information Tests

        [Fact]
        public void GetIpv4Addresses_ReturnsList()
        {
            List<string> addressesIncludeDown = MdlNet.GetIpv4Addresses(includeDownInterfaces: true);
            List<string> addressesActiveOnly = MdlNet.GetIpv4Addresses(includeDownInterfaces: false);

            Assert.NotNull(addressesIncludeDown);
            Assert.NotNull(addressesActiveOnly);

            foreach (var ipStr in addressesActiveOnly)
            {
                Assert.True(IPAddress.TryParse(ipStr, out var ip));
                Assert.Equal(AddressFamily.InterNetwork, ip.AddressFamily);
            }
        }

        #endregion

        #region Obsolete Methods Tests

#pragma warning disable CS0618

        [Fact]
        public void Obsolete_IPAddressToHostname()
        {
            string hostname = MdlNet.IPAddressToHostname("127.0.0.1");
            Assert.False(string.IsNullOrEmpty(hostname));
        }

        [Fact]
        public void Obsolete_HostnameToIPAddr()
        {
            IPAddress? ip = MdlNet.HostnameToIPAddr("localhost", isIPv6: false);
            Assert.NotNull(ip);
        }

        [Fact]
        public void Obsolete_HostnameToIP()
        {
            string? ipStr = MdlNet.HostnameToIP("localhost", isIPv6: false);
            Assert.NotNull(ipStr);
        }

        [Fact]
        public void Obsolete_GetIpv4List()
        {
            List<string> list = MdlNet.GetIpv4List(includeDownInterfaces: true);
            Assert.NotNull(list);
        }

#pragma warning restore CS0618

        #endregion
    }
}
